zapgrpc/zapgrpc.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package zapgrpc provides a logger that is compatible with grpclog.
+package zapgrpc // import "go.uber.org/zap/zapgrpc"
+
+import "go.uber.org/zap"
+
+// LoggerOption is an option for a new Logger.
+type LoggerOption func(*loggerOptions)
+
+// WithDebug says to print the debug level instead of the default info level.
+func WithDebug() LoggerOption {
+	return func(loggerOptions *loggerOptions) {
+		loggerOptions.printFunc = (*zap.SugaredLogger).Debug
+		loggerOptions.printfFunc = (*zap.SugaredLogger).Debugf
+	}
+}
+
+// NewLogger returns a new Logger.
+func NewLogger(l *zap.Logger, options ...LoggerOption) *Logger {
+	loggerOptions := newLoggerOptions(options...)
+	return &Logger{
+		l.Sugar(),
+		(*zap.SugaredLogger).Fatal,
+		(*zap.SugaredLogger).Fatalf,
+		loggerOptions.printFunc,
+		loggerOptions.printfFunc,
+	}
+}
+
+// Logger is a logger that is compatible with the grpclog Logger interface.
+type Logger struct {
+	sugaredLogger *zap.SugaredLogger
+	fatalFunc     func(*zap.SugaredLogger, ...interface{})
+	fatalfFunc    func(*zap.SugaredLogger, string, ...interface{})
+	printFunc     func(*zap.SugaredLogger, ...interface{})
+	printfFunc    func(*zap.SugaredLogger, string, ...interface{})
+}
```

Comment:
Maybe there's a misunderstanding - note this is controlling what happens when you call a `grpclog` function such as `grpclog.Printf`, not controlling the actual `*zap.Logger` given to `zapgrpc.SetLogger`, so for example:

```go
grpclog.Print("hello") # if WithDebug was not set, this will call zap.Info
grpclog.Print("hello") # if WithDebug was set, this will call zap.Debug
```

By setting `WithDebug`, if the level of the `*zap.Logger` is not debug, this will suppress grpc logging, which might be what someone wants.

Doing this by setting `printFunc, printfFunc` versus having a `debug` bool to me is cleaner, and the difference is minor, something around a few lines. I think the function style is nicer, and also note how it lets me test `grpclog.Fatal*` functions easily - without being able to also set `fatalFunc, fatalfFunc`, I wouldn't be really able to test `grpclog.Fatal*` functions because the program would exit. Without adding any special boolean to the actual implementation, I am able to add a `withWarn` `LoggerOption` in the tests that sets the fatal functions to redirect to the warn level, in the same style as the `WithDebug` option. This alone might be a good tie breaker.